<a href="https://colab.research.google.com/github/vanditasharma777/MNIST-Classification-using-CNN/blob/master/CNN_MNIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### COMPARISON OF CNN FOR VARIATIONS OF ACTIVATION FUNCTION                          AND NUMBER OF CONVOLUTIONAL LAYERS FOR MNIST DATASET 

In [0]:
import keras
import tensorflow
from keras.datasets import mnist 
from keras.models import Sequential 
from keras.layers import Dense, Dropout, Flatten 
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as k
import keras.utils 
import sklearn
from sklearn.metrics import classification_report,confusion_matrix
batch_size = 128
num_classes = 10
epochs = 12

In [0]:
# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()


11493376/11490434 [==============================] - 2s 0us/step


In [0]:
#print the new label
if k.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')


x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [0]:
# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)


### **1. Using relu activation function with kernel size(3,3) with 2 Convolutional Layers**

In [0]:
# Build the CNN model
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 142s 2ms/step - loss: 0.2702 - accuracy: 0.9172 - val_loss: 0.0568 - val_accuracy: 0.9817
Epoch 2/12
60000/60000 [==============================] - 142s 2ms/step - loss: 0.0909 - accuracy: 0.9733 - val_loss: 0.0413 - val_accuracy: 0.9857
Epoch 3/12
60000/60000 [==============================] - 141s 2ms/step - loss: 0.0666 - accuracy: 0.9808 - val_loss: 0.0338 - val_accuracy: 0.9888
Epoch 4/12
60000/60000 [==============================] - 141s 2ms/step - loss: 0.0558 - accuracy: 0.9830 - val_loss: 0.0306 - val_accuracy: 0.9896
Epoch 5/12
60000/60000 [==============================] - 141s 2ms/step - loss: 0.0475 - accuracy: 0.9858 - val_loss: 0.0291 - val_accuracy: 0.9898
Epoch 6/12
60000/60000 [==============================] - 142s 2ms/step - loss: 0.0424 - accuracy: 0.9872 - val_loss: 0.0320 - val_accuracy: 0.9887
Epoch 7/12
60000/60000 [==============================] - 141s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 975    0    1    0    0    0    1    0    2    1]
 [   0 1133    0    1    0    1    0    0    0    0]
 [   1    3 1019    1    1    0    0    5    2    0]
 [   0    0    1 1007    0    1    0    0    1    0]
 [   0    0    0    0  974    0    3    0    1    4]
 [   0    0    0    5    0  886    1    0    0    0]
 [   4    2    0    0    1    4  944    0    3    0]
 [   0    3    3    1    0    0    0 1019    1    1]
 [   1    0    1    2    0    0    0    1  967    2]
 [   0    1    0    1    4    4    0    3    4  992]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **2. Using tanh activation function with kernel_size (3,3) with 2 CNN Layers**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3), activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 145s 2ms/step - loss: 0.2486 - accuracy: 0.9247 - val_loss: 0.1004 - val_accuracy: 0.9662
Epoch 2/12
60000/60000 [==============================] - 145s 2ms/step - loss: 0.1012 - accuracy: 0.9699 - val_loss: 0.0657 - val_accuracy: 0.9790
Epoch 3/12
60000/60000 [==============================] - 145s 2ms/step - loss: 0.0787 - accuracy: 0.9759 - val_loss: 0.0516 - val_accuracy: 0.9828
Epoch 4/12
60000/60000 [==============================] - 145s 2ms/step - loss: 0.0625 - accuracy: 0.9804 - val_loss: 0.0468 - val_accuracy: 0.9844
Epoch 5/12
60000/60000 [==============================] - 144s 2ms/step - loss: 0.0561 - accuracy: 0.9823 - val_loss: 0.0407 - val_accuracy: 0.9873
Epoch 6/12
60000/60000 [==============================] - 145s 2ms/step - loss: 0.0494 - accuracy: 0.9847 - val_loss: 0.0387 - val_accuracy: 0.9881
Epoch 7/12
60000/60000 [==============================] - 145s

In [0]:
  test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    0    1    0    0    1    1    1    0    0]
 [   0 1130    1    1    0    0    1    1    1    0]
 [   1    1 1019    0    2    0    3    5    1    0]
 [   0    0    1  998    0    5    0    2    3    1]
 [   0    0    1    0  978    0    1    0    0    2]
 [   2    0    1    8    0  879    1    0    1    0]
 [   5    2    0    1    2    1  947    0    0    0]
 [   0    1    7    1    0    0    0 1018    1    0]
 [   2    2    3    1    1    1    0    2  959    3]
 [   1    2    0    1    9    4    0    3    2  987]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **3. Using relu activation function with kernel_size (4,4) and tanh activation function with Kernel_size (3,3)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])



Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 138s 2ms/step - loss: 0.2194 - accuracy: 0.9332 - val_loss: 0.0582 - val_accuracy: 0.9826
Epoch 2/12
60000/60000 [==============================] - 133s 2ms/step - loss: 0.0783 - accuracy: 0.9768 - val_loss: 0.0502 - val_accuracy: 0.9832
Epoch 3/12
60000/60000 [==============================] - 133s 2ms/step - loss: 0.0625 - accuracy: 0.9809 - val_loss: 0.0376 - val_accuracy: 0.9874
Epoch 4/12
60000/60000 [==============================] - 133s 2ms/step - loss: 0.0511 - accuracy: 0.9845 - val_loss: 0.0372 - val_accuracy: 0.9886
Epoch 5/12
60000/60000 [==============================] - 138s 2ms/step - loss: 0.0445 - accuracy: 0.9865 - val_loss: 0.0337 - val_accuracy: 0.9886
Epoch 6/12
60000/60000 [==============================] - 133s 2ms/step - loss: 0.0423 - accuracy: 0.9872 - val_loss: 0.0372 - val_accuracy: 0.9880
Epoch 7/12
60000/60000 [==============================] - 133s

In [0]:

test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 975    0    2    0    0    0    1    1    1    0]
 [   0 1131    1    1    0    0    1    1    0    0]
 [   2    1 1024    1    1    0    0    3    0    0]
 [   0    0    3 1005    0    1    0    0    1    0]
 [   1    0    1    0  970    0    4    0    0    6]
 [   2    0    1    5    0  882    1    0    0    1]
 [   4    1    0    1    3    3  946    0    0    0]
 [   0    1    8    1    0    0    0 1014    1    3]
 [   5    0    2    2    1    0    0    1  961    2]
 [   0    2    0    1    3    3    0    4    1  995]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **4. Using relu activation function with kernel_size (3,3) and tanh activation function with Kernel_size (4,4)**


In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 201s 3ms/step - loss: 0.2555 - accuracy: 0.9224 - val_loss: 0.0670 - val_accuracy: 0.9799
Epoch 2/12
60000/60000 [==============================] - 201s 3ms/step - loss: 0.0921 - accuracy: 0.9717 - val_loss: 0.0411 - val_accuracy: 0.9866
Epoch 3/12
60000/60000 [==============================] - 200s 3ms/step - loss: 0.0679 - accuracy: 0.9796 - val_loss: 0.0376 - val_accuracy: 0.9876
Epoch 4/12
60000/60000 [==============================] - 201s 3ms/step - loss: 0.0569 - accuracy: 0.9826 - val_loss: 0.0350 - val_accuracy: 0.9885
Epoch 5/12
60000/60000 [==============================] - 200s 3ms/step - loss: 0.0488 - accuracy: 0.9853 - val_loss: 0.0279 - val_accuracy: 0.9905
Epoch 6/12
60000/60000 [==============================] - 201s 3ms/step - loss: 0.0450 - accuracy: 0.9863 - val_loss: 0.0289 - val_accuracy: 0.9901
Epoch 7/12
60000/60000 [==============================] - 205s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    0    1    0    0    0    1    0    0    1]
 [   0 1132    2    0    0    1    0    0    0    0]
 [   1    1 1028    0    0    0    0    2    0    0]
 [   0    0    3 1003    0    3    0    0    1    0]
 [   0    0    1    0  977    0    2    0    0    2]
 [   0    0    0    7    0  884    1    0    0    0]
 [   4    2    0    0    1    4  947    0    0    0]
 [   0    2    3    0    0    0    0 1021    1    1]
 [   2    0    2    1    0    1    0    1  965    2]
 [   1    1    0    0    3    4    0    3    2  995]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **5. Using tanh activation function with kernel_size (4,4) with 2 CNN layers**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4), activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 190s 3ms/step - loss: 0.2361 - accuracy: 0.9284 - val_loss: 0.0690 - val_accuracy: 0.9789
Epoch 2/12
60000/60000 [==============================] - 191s 3ms/step - loss: 0.0961 - accuracy: 0.9713 - val_loss: 0.0511 - val_accuracy: 0.9835
Epoch 3/12
60000/60000 [==============================] - 190s 3ms/step - loss: 0.0741 - accuracy: 0.9777 - val_loss: 0.0435 - val_accuracy: 0.9863
Epoch 4/12
60000/60000 [==============================] - 190s 3ms/step - loss: 0.0609 - accuracy: 0.9813 - val_loss: 0.0374 - val_accuracy: 0.9866
Epoch 5/12
60000/60000 [==============================] - 190s 3ms/step - loss: 0.0540 - accuracy: 0.9834 - val_loss: 0.0378 - val_accuracy: 0.9864
Epoch 6/12
60000/60000 [==============================] - 190s 3ms/step - loss: 0.0465 - accuracy: 0.9855 - val_loss: 0.0352 - val_accuracy: 0.9875
Epoch 7/12
60000/60000 [==============================] - 190s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    0    1    0    0    0    1    1    1    0]
 [   0 1125    1    2    0    0    5    1    1    0]
 [   3    1 1018    0    1    0    1    4    4    0]
 [   0    0    1 1003    0    3    0    0    3    0]
 [   0    0    0    0  972    0    2    0    1    7]
 [   1    0    1    7    0  881    2    0    0    0]
 [   4    1    0    1    1    2  945    0    4    0]
 [   0    2    7    0    0    0    0 1015    1    3]
 [   2    0    1    1    0    0    0    1  968    1]
 [   3    1    0    3    5    4    0    1    5  987]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **6. Using relu activation function with kernel_size (4,4) with 2 CNN layers**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 189s 3ms/step - loss: 0.2398 - accuracy: 0.9268 - val_loss: 0.0553 - val_accuracy: 0.9817
Epoch 2/12
60000/60000 [==============================] - 188s 3ms/step - loss: 0.0797 - accuracy: 0.9763 - val_loss: 0.0349 - val_accuracy: 0.9882
Epoch 3/12
60000/60000 [==============================] - 188s 3ms/step - loss: 0.0604 - accuracy: 0.9822 - val_loss: 0.0305 - val_accuracy: 0.9888
Epoch 4/12
60000/60000 [==============================] - 188s 3ms/step - loss: 0.0499 - accuracy: 0.9855 - val_loss: 0.0294 - val_accuracy: 0.9900
Epoch 5/12
60000/60000 [==============================] - 187s 3ms/step - loss: 0.0437 - accuracy: 0.9876 - val_loss: 0.0256 - val_accuracy: 0.9908
Epoch 6/12
60000/60000 [==============================] - 188s 3ms/step - loss: 0.0378 - accuracy: 0.9886 - val_loss: 0.0252 - val_accuracy: 0.9903
Epoch 7/12
60000/60000 [==============================] - 188s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm) 

[[ 977    0    1    0    0    0    1    1    0    0]
 [   0 1134    0    0    0    0    1    0    0    0]
 [   1    2 1024    0    0    0    0    5    0    0]
 [   0    0    2 1005    0    2    0    0    1    0]
 [   0    0    0    0  978    0    2    0    0    2]
 [   2    0    0    5    0  884    1    0    0    0]
 [   4    2    0    0    1    4  947    0    0    0]
 [   0    1    3    2    0    0    0 1020    1    1]
 [   4    0    2    2    0    1    0    0  964    1]
 [   0    0    0    1    4    3    0    1    1  999]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **7. Using relu activation function with kernel_size(5,5)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(5, 5),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
  print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 226s 4ms/step - loss: 0.1817 - accuracy: 0.9446 - val_loss: 0.0533 - val_accuracy: 0.9830
Epoch 2/12
60000/60000 [==============================] - 226s 4ms/step - loss: 0.0613 - accuracy: 0.9813 - val_loss: 0.0343 - val_accuracy: 0.9887
Epoch 3/12
60000/60000 [==============================] - 226s 4ms/step - loss: 0.0473 - accuracy: 0.9864 - val_loss: 0.0311 - val_accuracy: 0.9883
Epoch 4/12
60000/60000 [==============================] - 227s 4ms/step - loss: 0.0378 - accuracy: 0.9884 - val_loss: 0.0298 - val_accuracy: 0.9906
Epoch 5/12
60000/60000 [==============================] - 227s 4ms/step - loss: 0.0329 - accuracy: 0.9896 - val_loss: 0.0235 - val_accuracy: 0.9923
Epoch 6/12
60000/60000 [==============================] - 227s 4ms/step - loss: 0.0280 - accuracy: 0.9914 - val_loss: 0.0232 - val_accuracy: 0.9913
Epoch 7/12
60000/60000 [==============================] - 227s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    0    1    0    0    0    0    1    1    0]
 [   1 1128    2    1    0    0    1    0    2    0]
 [   1    0 1028    0    2    0    0    1    0    0]
 [   0    0    1 1003    0    2    0    0    4    0]
 [   0    0    0    0  976    0    0    0    2    4]
 [   1    0    0    6    0  883    1    0    1    0]
 [   4    2    0    1    2    1  944    0    4    0]
 [   0    2    2    1    0    0    0 1021    1    1]
 [   0    0    1    0    0    0    0    0  972    1]
 [   0    0    0    1    3    1    0    1    6  997]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **8. Using tanh activation function with kernel_size(5,5)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(5, 5),
                 activation='tanh',
                 input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 229s 4ms/step - loss: 0.2418 - accuracy: 0.9262 - val_loss: 0.0737 - val_accuracy: 0.9769
Epoch 2/12
60000/60000 [==============================] - 234s 4ms/step - loss: 0.1038 - accuracy: 0.9690 - val_loss: 0.0455 - val_accuracy: 0.9843
Epoch 3/12
60000/60000 [==============================] - 230s 4ms/step - loss: 0.0781 - accuracy: 0.9765 - val_loss: 0.0408 - val_accuracy: 0.9854
Epoch 4/12
60000/60000 [==============================] - 230s 4ms/step - loss: 0.0615 - accuracy: 0.9814 - val_loss: 0.0375 - val_accuracy: 0.9871
Epoch 5/12
60000/60000 [==============================] - 237s 4ms/step - loss: 0.0555 - accuracy: 0.9830 - val_loss: 0.0335 - val_accuracy: 0.9882
Epoch 6/12
60000/60000 [==============================] - 231s 4ms/step - loss: 0.0499 - accuracy: 0.9847 - val_loss: 0.0344 - val_accuracy: 0.9877
Epoch 7/12
60000/60000 [==============================] - 231s

In [0]:
  test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 974    0    3    0    0    0    1    1    1    0]
 [   0 1131    0    0    0    0    2    1    1    0]
 [   0    1 1026    0    1    0    0    3    1    0]
 [   0    0    1 1005    0    2    0    0    2    0]
 [   0    0    0    0  978    0    3    0    0    1]
 [   0    0    0    6    0  884    2    0    0    0]
 [   4    1    0    0    1    1  951    0    0    0]
 [   0    2    4    3    0    0    0 1018    1    0]
 [   4    0    2    1    1    1    1    1  962    1]
 [   0    2    2    1   10    7    0    8    5  974]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **9. Using 3 CNN layers with activation function tanh and Kernel_size(3,3)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='tanh',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 284s 5ms/step - loss: 0.2241 - accuracy: 0.9305 - val_loss: 0.0685 - val_accuracy: 0.9793
Epoch 2/12
60000/60000 [==============================] - 280s 5ms/step - loss: 0.0988 - accuracy: 0.9702 - val_loss: 0.0496 - val_accuracy: 0.9839
Epoch 3/12
60000/60000 [==============================] - 285s 5ms/step - loss: 0.0719 - accuracy: 0.9788 - val_loss: 0.0430 - val_accuracy: 0.9859
Epoch 4/12
60000/60000 [==============================] - 288s 5ms/step - loss: 0.0603 - accuracy: 0.9818 - val_loss: 0.0452 - val_accuracy: 0.9847
Epoch 5/12
60000/60000 [==============================] - 284s 5ms/step - loss: 0.0535 - accuracy: 0.9833 - val_loss: 0.0406 - val_accuracy: 0.9863
Epoch 6/12
60000/60000 [==============================] - 289s 5ms/step - loss: 0.0450 - accuracy: 0.9858 - val_loss: 0.0353 - val_accuracy: 0.9885
Epoch 7/12
60000/60000 [==============================] - 286s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    0    1    0    0    0    1    1    0    0]
 [   0 1134    0    0    0    0    1    0    0    0]
 [   1    2 1024    0    0    0    0    5    0    0]
 [   0    0    2 1005    0    2    0    0    1    0]
 [   0    0    0    0  978    0    2    0    0    2]
 [   2    0    0    5    0  884    1    0    0    0]
 [   4    2    0    0    1    4  947    0    0    0]
 [   0    1    3    2    0    0    0 1020    1    1]
 [   4    0    2    2    0    1    0    0  964    1]
 [   0    0    0    1    4    3    0    1    1  999]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **10. Using 3 CNN layers with activation function relu and Kernel_size(3,3)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='relu', input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.2430 - accuracy: 0.9249 - val_loss: 0.0492 - val_accuracy: 0.9843
Epoch 2/12
60000/60000 [==============================] - 301s 5ms/step - loss: 0.0799 - accuracy: 0.9767 - val_loss: 0.0395 - val_accuracy: 0.9870
Epoch 3/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0593 - accuracy: 0.9823 - val_loss: 0.0281 - val_accuracy: 0.9903
Epoch 4/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0499 - accuracy: 0.9855 - val_loss: 0.0293 - val_accuracy: 0.9904
Epoch 5/12
60000/60000 [==============================] - 301s 5ms/step - loss: 0.0423 - accuracy: 0.9873 - val_loss: 0.0252 - val_accuracy: 0.9912
Epoch 6/12
60000/60000 [==============================] - 301s 5ms/step - loss: 0.0395 - accuracy: 0.9881 - val_loss: 0.0246 - val_accuracy: 0.9918
Epoch 7/12
60000/60000 [==============================] - 301s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    0    1    0    0    0    1    1    0    0]
 [   0 1134    0    0    0    0    1    0    0    0]
 [   1    2 1024    0    0    0    0    5    0    0]
 [   0    0    2 1005    0    2    0    0    1    0]
 [   0    0    0    0  978    0    2    0    0    2]
 [   2    0    0    5    0  884    1    0    0    0]
 [   4    2    0    0    1    4  947    0    0    0]
 [   0    1    3    2    0    0    0 1020    1    1]
 [   4    0    2    2    0    1    0    0  964    1]
 [   0    0    0    1    4    3    0    1    1  999]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **11. With single layer of dropout with 3 CNN layers and kernel_size(3,3)**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 303s 5ms/step - loss: 0.2190 - accuracy: 0.9324 - val_loss: 0.0614 - val_accuracy: 0.9797
Epoch 2/12
60000/60000 [==============================] - 309s 5ms/step - loss: 0.0888 - accuracy: 0.9733 - val_loss: 0.0569 - val_accuracy: 0.9815
Epoch 3/12
60000/60000 [==============================] - 302s 5ms/step - loss: 0.0677 - accuracy: 0.9794 - val_loss: 0.0406 - val_accuracy: 0.9874
Epoch 4/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0549 - accuracy: 0.9834 - val_loss: 0.0408 - val_accuracy: 0.9873
Epoch 5/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0470 - accuracy: 0.9857 - val_loss: 0.0361 - val_accuracy: 0.9878
Epoch 6/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0416 - accuracy: 0.9867 - val_loss: 0.0358 - val_accuracy: 0.9872
Epoch 7/12
60000/60000 [==============================] - 300s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    0    1    0    0    0    1    1    1    0]
 [   0 1128    1    2    1    0    3    0    0    0]
 [   2    2 1022    0    1    0    1    4    0    0]
 [   0    0    1 1004    0    3    0    0    2    0]
 [   0    0    0    0  976    0    4    0    0    2]
 [   2    0    0    4    0  882    3    0    1    0]
 [   6    1    0    0    1    2  948    0    0    0]
 [   1    2    4    2    0    0    0 1017    1    1]
 [   5    1    3    1    0    0    1    1  961    1]
 [   3    2    0    1    6    4    0    3    2  988]]


In [0]:
acc = accuracy_score(rounded_labels,test_predictions)
print(acc)

0.9902


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **12. Using tanh activation function Without Dropout Layer**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.1613 - accuracy: 0.9510 - val_loss: 0.0576 - val_accuracy: 0.9814
Epoch 2/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0467 - accuracy: 0.9869 - val_loss: 0.0446 - val_accuracy: 0.9853
Epoch 3/12
60000/60000 [==============================] - 305s 5ms/step - loss: 0.0270 - accuracy: 0.9923 - val_loss: 0.0398 - val_accuracy: 0.9874
Epoch 4/12
60000/60000 [==============================] - 301s 5ms/step - loss: 0.0155 - accuracy: 0.9958 - val_loss: 0.0349 - val_accuracy: 0.9898
Epoch 5/12
60000/60000 [==============================] - 301s 5ms/step - loss: 0.0080 - accuracy: 0.9985 - val_loss: 0.0306 - val_accuracy: 0.9905
Epoch 6/12
60000/60000 [==============================] - 304s 5ms/step - loss: 0.0037 - accuracy: 0.9995 - val_loss: 0.0292 - val_accuracy: 0.9906
Epoch 7/12
60000/60000 [==============================] - 300s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    0    0    0    0    0    1    1    2    0]
 [   0 1128    1    1    0    0    3    1    1    0]
 [   1    2 1023    0    1    0    0    4    1    0]
 [   0    0    2 1005    0    2    0    0    1    0]
 [   0    0    0    0  975    0    4    0    0    3]
 [   1    0    0    5    0  885    1    0    0    0]
 [   5    2    0    1    1    3  945    0    1    0]
 [   0    1    5    1    0    0    0 1020    1    0]
 [   3    0    3    1    0    0    0    1  965    1]
 [   0    1    0    2    4    1    1    3    1  996]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **13. Using relu activation function without dropout layer**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dense(num_classes, activation='softmax'))
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 296s 5ms/step - loss: 0.1593 - accuracy: 0.9516 - val_loss: 0.0523 - val_accuracy: 0.9833
Epoch 2/12
60000/60000 [==============================] - 297s 5ms/step - loss: 0.0398 - accuracy: 0.9880 - val_loss: 0.0365 - val_accuracy: 0.9878
Epoch 3/12
60000/60000 [==============================] - 296s 5ms/step - loss: 0.0237 - accuracy: 0.9930 - val_loss: 0.0347 - val_accuracy: 0.9892
Epoch 4/12
60000/60000 [==============================] - 296s 5ms/step - loss: 0.0150 - accuracy: 0.9958 - val_loss: 0.0305 - val_accuracy: 0.9895
Epoch 5/12
60000/60000 [==============================] - 295s 5ms/step - loss: 0.0093 - accuracy: 0.9977 - val_loss: 0.0270 - val_accuracy: 0.9915
Epoch 6/12
60000/60000 [==============================] - 294s 5ms/step - loss: 0.0055 - accuracy: 0.9988 - val_loss: 0.0264 - val_accuracy: 0.9922
Epoch 7/12
60000/60000 [==============================] - 294s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    0    0    0    0    0    1    1    2    0]
 [   0 1133    0    0    0    0    1    1    0    0]
 [   1    1 1027    0    1    0    0    1    1    0]
 [   0    0    2 1005    0    2    0    0    1    0]
 [   0    0    1    0  975    0    2    0    1    3]
 [   2    0    0    6    0  882    2    0    0    0]
 [   5    2    0    0    1    1  948    0    1    0]
 [   0    1    7    0    1    0    0 1016    1    2]
 [   2    0    2    1    1    0    0    1  966    1]
 [   0    0    0    0    4    3    0    3    3  996]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **14.Using tanh, relu and tanh activation function with kernel size(4,4) and (5,5), (5,5) without dropout layer**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='relu'))
model.add(Conv2D(64, (5, 5), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dense(num_classes, activation='softmax'))
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 506s 8ms/step - loss: 0.1516 - accuracy: 0.9533 - val_loss: 0.0585 - val_accuracy: 0.9825
Epoch 2/12
60000/60000 [==============================] - 506s 8ms/step - loss: 0.0414 - accuracy: 0.9877 - val_loss: 0.0397 - val_accuracy: 0.9871
Epoch 3/12
60000/60000 [==============================] - 507s 8ms/step - loss: 0.0257 - accuracy: 0.9924 - val_loss: 0.0280 - val_accuracy: 0.9907
Epoch 4/12
60000/60000 [==============================] - 506s 8ms/step - loss: 0.0166 - accuracy: 0.9951 - val_loss: 0.0267 - val_accuracy: 0.9910
Epoch 5/12
60000/60000 [==============================] - 505s 8ms/step - loss: 0.0103 - accuracy: 0.9972 - val_loss: 0.0267 - val_accuracy: 0.9906
Epoch 6/12
60000/60000 [==============================] - 507s 8ms/step - loss: 0.0058 - accuracy: 0.9988 - val_loss: 0.0261 - val_accuracy: 0.9910
Epoch 7/12
60000/60000 [==============================] - 506s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    1    0    0    0    0    1    0    1    0]
 [   0 1131    0    1    1    0    1    0    1    0]
 [   2    2 1026    0    0    0    0    2    0    0]
 [   0    0    1 1005    0    2    0    0    2    0]
 [   0    0    1    0  977    0    3    0    0    1]
 [   1    0    0    6    0  883    1    0    1    0]
 [   4    2    0    1    1    2  946    0    2    0]
 [   0    2    2    0    0    0    0 1022    1    1]
 [   3    0    1    1    0    1    0    1  966    1]
 [   0    1    0    0    5    3    0    1    2  997]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **15. Using 5 CNN layers**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 471s 8ms/step - loss: 0.2366 - accuracy: 0.9277 - val_loss: 0.0899 - val_accuracy: 0.9721
Epoch 2/12
60000/60000 [==============================] - 472s 8ms/step - loss: 0.0913 - accuracy: 0.9726 - val_loss: 0.0435 - val_accuracy: 0.9853
Epoch 3/12
60000/60000 [==============================] - 471s 8ms/step - loss: 0.0716 - accuracy: 0.9776 - val_loss: 0.0386 - val_accuracy: 0.9870
Epoch 4/12
60000/60000 [==============================] - 472s 8ms/step - loss: 0.0589 - accuracy: 0.9819 - val_loss: 0.0320 - val_accuracy: 0.9897
Epoch 5/12
60000/60000 [==============================] - 468s 8ms/step - loss: 0.0523 - accuracy: 0.9839 - val_loss: 0.0317 - val_accuracy: 0.9894
Epoch 6/12
60000/60000 [==============================] - 476s 8ms/step - loss: 0.0446 - accuracy: 0.9860 - val_loss: 0.0297 - val_accuracy: 0.9906
Epoch 7/12
60000/60000 [==============================] - 478s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 976    1    1    0    0    0    1    1    0    0]
 [   0 1135    0    0    0    0    0    0    0    0]
 [   1    7 1015    2    1    0    1    5    0    0]
 [   0    0    1  998    0    6    0    3    1    1]
 [   1    1    0    0  969    0    5    0    1    5]
 [   2    0    0    2    0  885    2    1    0    0]
 [   4    4    0    1    0    2  947    0    0    0]
 [   0    2    1    1    0    0    0 1023    1    0]
 [   3    1    4    1    2    3    2    2  953    3]
 [   2    4    0    0    5    3    0    2    2  991]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **16. By changing the activation function in various layers**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
 
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 472s 8ms/step - loss: 0.1894 - accuracy: 0.9416 - val_loss: 0.0458 - val_accuracy: 0.9849
Epoch 2/12
60000/60000 [==============================] - 471s 8ms/step - loss: 0.0657 - accuracy: 0.9801 - val_loss: 0.0306 - val_accuracy: 0.9900
Epoch 3/12
60000/60000 [==============================] - 472s 8ms/step - loss: 0.0513 - accuracy: 0.9845 - val_loss: 0.0343 - val_accuracy: 0.9892
Epoch 4/12
60000/60000 [==============================] - 471s 8ms/step - loss: 0.0431 - accuracy: 0.9871 - val_loss: 0.0275 - val_accuracy: 0.9911
Epoch 5/12
60000/60000 [==============================] - 471s 8ms/step - loss: 0.0369 - accuracy: 0.9886 - val_loss: 0.0225 - val_accuracy: 0.9929
Epoch 6/12
60000/60000 [==============================] - 468s 8ms/step - loss: 0.0318 - accuracy: 0.9900 - val_loss: 0.0251 - val_accuracy: 0.9920
Epoch 7/12
60000/60000 [==============================] - 476s

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
print("Confusion_matrix:")
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)


Confusion_matrix:
[[ 977    1    0    0    0    0    1    0    1    0]
 [   0 1131    0    1    1    0    1    0    1    0]
 [   2    2 1026    0    0    0    0    2    0    0]
 [   0    0    1 1005    0    2    0    0    2    0]
 [   0    0    1    0  977    0    3    0    0    1]
 [   1    0    0    6    0  883    1    0    1    0]
 [   4    2    0    1    1    2  946    0    2    0]
 [   0    2    2    0    0    0    0 1022    1    1]
 [   3    0    1    1    0    1    0    1  966    1]
 [   0    1    0    0    5    3    0    1    2  997]]


In [0]:
print("classification_report:")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



### **17. By changing the architecture**

In [0]:
model = keras.models.Sequential()
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(96, (3, 3), activation='tanh'))
model.add(Conv2D(128, (3, 3), activation='tanh'))
model.add(Conv2D(96, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='tanh'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=batch_size,epochs=epochs,verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 802s 13ms/step - loss: 0.2897 - accuracy: 0.9131 - val_loss: 0.0841 - val_accuracy: 0.9737
Epoch 2/12
60000/60000 [==============================] - 793s 13ms/step - loss: 0.1082 - accuracy: 0.9668 - val_loss: 0.0540 - val_accuracy: 0.9826
Epoch 3/12
60000/60000 [==============================] - 790s 13ms/step - loss: 0.0851 - accuracy: 0.9746 - val_loss: 0.0508 - val_accuracy: 0.9832
Epoch 4/12
60000/60000 [==============================] - 801s 13ms/step - loss: 0.0696 - accuracy: 0.9790 - val_loss: 0.0403 - val_accuracy: 0.9854
Epoch 5/12
60000/60000 [==============================] - 802s 13ms/step - loss: 0.0623 - accuracy: 0.9815 - val_loss: 0.0355 - val_accuracy: 0.9876
Epoch 6/12
60000/60000 [==============================] - 786s 13ms/step - loss: 0.0559 - accuracy: 0.9831 - val_loss: 0.0357 - val_accuracy: 0.9884
Epoch 7/12
60000/60000 [==============================] 

In [0]:
test_predictions = model.predict_classes(x_test, batch_size=128, verbose=0) 
import numpy as np
rounded_labels=np.argmax(y_test, axis=1)
rounded_labels[1]
cm = confusion_matrix(rounded_labels, test_predictions)
print(cm)

[[ 977    1    0    0    0    0    1    0    1    0]
 [   0 1131    0    1    1    0    1    0    1    0]
 [   2    2 1026    0    0    0    0    2    0    0]
 [   0    0    1 1005    0    2    0    0    2    0]
 [   0    0    1    0  977    0    3    0    0    1]
 [   1    0    0    6    0  883    1    0    1    0]
 [   4    2    0    1    1    2  946    0    2    0]
 [   0    2    2    0    0    0    0 1022    1    1]
 [   3    0    1    1    0    1    0    1  966    1]
 [   0    1    0    0    5    3    0    1    2  997]]


In [0]:
print("classification_report")
print(classification_report(rounded_labels, test_predictions, target_names= None))

classification_report
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       1.00      0.99      0.99      1032
           3       0.99      1.00      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.99      0.99       892
           6       0.99      0.99      0.99       958
           7       1.00      0.99      1.00      1028
           8       0.99      0.99      0.99       974
           9       1.00      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



# **RESULTS AND CONCLUSION**
1.  Activation Function=relu,relu, No. of CNN layers=2, Kernel-size=(3,3),(3,3) , Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0273, Test Accuracy=0.9916 
2.  Activation Function=tanh,tanh, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(3,3),(3,3), Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0366 Test Accuracy=0.9890 
3.  Activation Function=relu,tanh, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(4,4), (3,3), Dropout Layers=2, Dropout Rate=0.25,0.5,Test loss=0.0292 Test Accuracy=0.9903 
4.  Activation Function=relu,tanh, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(3,3),(4,4), Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0287 Test Accuracy=0.9919 
5.  Activation Function=tanh,tanh, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(4,4), (4,4), Dropout Layers=2, Dropout Rate=0.25,0.5,Test loss=0.0362 Test Accuracy=0.9890 
6.  Activation Function=relu,relu, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(4,4), (4,4), Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0223 Test Accuracy=0.9929 
7.  Activation Function=relu,relu, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(5,5),(5,5), Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0226 Test Accuracy=0.9929 
8.  Activation Function=tanh,tanh, No. of CNN layers=2, Kernel-size=(3,3),(3,3)(5,5),(5,5), Dropout Layers=2, Dropout Rate=0.25,0.25, Test loss=0.0272 Test Accuracy=0.9902 
9.  Activation Function=tanh, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(3,3), (3,3), (3,3), Dropout Layers=2, Dropout Rate=0.25,0.5,Test loss=0.0366 Test Accuracy=0.9887 
10.  Activation Function=relu, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(3,3), (3,3), (3,3), Dropout Layers=2,  Dropout Rate=0.25,0.5, Test loss=0.0218,Test Accuracy= 0.9932 
11.  Activation Function=tanh, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(3,3), (3,3), (3,3), Dropout Layers=1, Dropout Rate=0.25, Test loss=0.0309, Test Accuracy=0.9901 
12.  Activation Function=tanh, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(3,3), (3,3), (3,3), Dropout Layers=0, Dropout Rate=NA, Test loss=0.0304, Test Accuracy=0.9918 
13.  Activation Function=relu, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(3,3), (3,3), (3,3), Dropout Layers=0, Dropout Rate=NA,Test loss=0.0276,Test Accuracy=0.9923 
14.  Activation Function=tanh,relu,tanh, No. of CNN layers=3, Kernel-size=(3,3),(3,3)(4,4),(5,5),(5,5), Dropout Layers=0, Dropout Rate=NA,Test loss=0.0237 0.9929 
15.  Activation Function=tanh, No. of CNN layers=5, Kernel-size=(3,3),(3,3)(3,3) for each layer, Dropout Layers=2, Dropout Rate=0.25,0.5,Test loss=0.0332,Test Accuracy=0.9891 
16.  Activation Function=tanh,relu, No. of CNN layers=5, Kernel-size=(3,3),(3,3)(3,3) for each layer, Dropout Layers=2, Dropout Rate=0.25,0.5,Test loss=0.0191,Test Accuracy=0.9940 
17.  Activation Function=tanh,relu, No. of CNN layers=5, Kernel-size=(3,3),(3,3)(3,3) for each layer, Dropout Layers=2, Dropout Rate=0.25,0.5, Test loss=0.0294,Test Accuracy=0.9922 
 
After comparing different CNN models for MNIST dataset, the highest accuracy is 99.4%, which is obtained by using five CNN layers with kernel-size (3, 3) for each layer, two dropout layers with dropout rate 0.25, 0.5 respectively. The evaluation metrics such as precision, recall and f1-score are not compare because MNIST is very simple dataset are almost same for all models. 
                                                                    